### PR TITLE
Ocelots - Dalia Ali

### PR DESCRIPTION
**Pull request by dnabilali**
completed viewing party assignment

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -118,13 +118,14 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     # Assert
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
+    assert updated_data["watched"][0]["title"] == MOVIE_TITLE_1
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:121]: Nice assertion! In this case where the code is checking individual keys of a dictionary, I recommend including assertions for all of the relevant keys.  
```python
    watched_movie = updated_data["watched"][0]
    assert watched_movie["title”] == MOVIE_TITLE_1
    assert watched_movie["rating"] == RATING_1
    assert watched_movie["genre"] == GENRE_1
```

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -142,13 +143,14 @@ def test_moves_movie_from_watchlist_to_watched():
     # Assert
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
+    assert updated_data["watched"][1]["title"] == "It Came from the Stack Trace"
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:146]: The README isn't explicit about what order we should add movies to the list `watched`. If we want our tests to be independent from a specific implementation that adds movies to a particular location in the `watched` list, we could use the `in` keyword to check if the variable `movie_to_watch` is in the `watched` list instead of looking at a specific index. By checking for the whole dictionary, there's a side benefit that we don't need to check each key independently:
```python
assert movie_to_watch in updated_data["watched"]

# Another option:
assert HORROR_1 in updated_data["watched"]
```

File: tests/test_wave_03.py [hunk 3]
```diff
@@ -54,13 +54,16 @@ def test_friends_unique_movies_not_duplicated():
 
     # Assert
     assert len(friends_unique_movies) == 3
-
-    raise Exception("Test needs to be completed.")
+    assert INTRIGUE_3 in friends_unique_movies
+    assert HORROR_1 in friends_unique_movies
+    assert FANTASY_4 in friends_unique_movies
```
Comment (kelsey-steven-ada) [inline tests/test_wave_03.py:59]: Great checks for all the relevant data!

File: tests/test_wave_05.py [hunk 4]
```diff
@@ -53,12 +53,18 @@ def test_new_genre_rec_from_empty_friends():
         ]
     }
 
-    raise Exception("Test needs to be completed.")
+    #Act
+    recommendations = get_new_rec_by_genre(sonyas_data)
+
+    #Assert
+    assert len(recommendations) == 0
```
Comment (kelsey-steven-ada) [inline tests/test_wave_05.py:60]: Great act & assert steps!

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,146 @@
 # ------------- WAVE 1 --------------------
 
-def create_movie(title, genre, rating):
-    pass
+def create_movie(movie_title, genre, rating):
+    movies = {}
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:4]: I suggest moving the `movies` declaration under the invalid data check so that we don't allocate space for `movies` unless we know it will be used. 

I would also suggest a singular name like `movie` for this variable. A plural name in programming generally implies that the variable holds a collection of several items. Since this variable holds the data for a single movie, a singular name would better represent what it holds.

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,146 @@
 # ------------- WAVE 1 --------------------
 
-def create_movie(title, genre, rating):
-    pass
+def create_movie(movie_title, genre, rating):
+    movies = {}
+
+    if movie_title == None or genre == None or rating == None :
+        return None
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:7]: Nice handling for returning `None` for missing data! In Python, when we test if a variable is pointing to `None` the PEP8 style guide recommends using `is` or `is not` for comparison rather than operators like `==`. 

`None` is a special object in Python, it’s a `singleton`, meaning there is only ever one instance of `None` in memory. All variables holding `None` are actually pointing to the same instance. By using `is` to compare against `None`, we avoid issues that could be caused by a class’s custom implementation of the comparison operators, and we gain some speed over a check using `==` since we are checking against identity rather than looking up if the variable's held type has a comparison function that should be used.
```python
if title is None or genre is None or rating is None :
        return None
```
If you’d like more info on `is` vs `==` for `None`, here’s a couple resources to get started =]
https://peps.python.org/pep-0008/#programming-recommendations
http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html


Something else to consider: here we're explicitly checking if the values are None, but what if we got an empty string for `title`? If we wanted to check for empty strings at the same time as we check for None, we could rely on the truthy/falsy values of the parameters and write something like: 
```python
if not title or not genre or not rating:
    return None
```

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,146 @@
...
+    movies = {}
+
+    if movie_title == None or genre == None or rating == None :
+        return None
+
+    movies["title"] = movie_title
+    movies["genre"] = genre
+    movies["rating"] = rating
+
+    return movies
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:13]: Nice approach! Another option could be to declare the values when we create `movies`:
```python
movies = {
    "title": title ,
    "genre": genre ,
    "rating": rating
}
return movies
```

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,146 @@
...
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def watch_movie(user_data, movie_title):
+    for movie in user_data["watchlist"]:
+        if movie_title in movie.values():
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:28]: `movie_title in movie.values()` will iterate over all the values in the `movie` dictionary. If we were in a system where it was important to reduce processing time, I'd suggest directly accessing the title key:
```python
if movie["title"] == movie_title:
```

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,146 @@
...
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+
+    return user_data
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:32]: This function works, but as an aside it contains an "anti-pattern" that is important to be aware of.  An element of the list that is being iterated over is being removed inside the for loop.  This won't cause problems in this particular case, but consider this slightly different implementation:
```
def watch_movie(user_data, title):
    for i in range(len(user_data["watchlist"])):
        item = user_data["watchlist"][i]

        if item["title"] == title:
            user_data["watched"].append(item)
            user_data["watchlist"].remove(item)

    return user_data
```
If we were to test this in a scenario where `user_data["watchlist"]` has 2 elements, and the first is the movie that we are moving to `user_data["watched"]`, the second iteration through the loop will throw an index error, because it will try to access the 2nd element in `user_data["watchlist"]`, but there no longer is a second element.  

As a coding pattern, this is one that I recommend avoiding because it has so much potential for unforeseen errors. One fix is to return immediately after making the swap (ie, add `return user_data` inside the if-block so that the loop ends as soon as the swap is made).  Another fix is to use the loop to find the item, but not make the swap until after the loop ends.

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,146 @@
...
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def  get_watched_avg_rating(user_data):
+    if len(user_data["watched"]) == 0:
+        return 0
+
+    ratings_sum = 0
+    movies_num = 0
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:43]: We could remove this variable and call `len(user_data["watched"])` where `movies_num` is used on line 49.

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,146 @@
...
+    
+    for movie in user_data["watched"]:
+        movies_num += 1
+        ratings_sum += movie["rating"]
+
+    average_rating = ratings_sum/movies_num
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:49]: The PEP8 guide on whitespace recommends we always place a single space on either side of an operator. This makes it easier for folks to visually process a line of code and see the groups of related statements in an expression. If you'd like more info, I suggest checking out the style guide for whitespace in expressions and statements: https://peps.python.org/pep-0008/#whitespace-in-expressions-and-statements

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,146 @@
...
+def  get_watched_avg_rating(user_data):
+    if len(user_data["watched"]) == 0:
+        return 0
+
+    ratings_sum = 0
+    movies_num = 0
+    
+    for movie in user_data["watched"]:
+        movies_num += 1
+        ratings_sum += movie["rating"]
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:47]: Nice approach! Another way we could do this is using list comprehensions and the `sum` function:

```python
# The list comprehension inside the `sum` call pulls the value of the 
# `rating` key out of each item in the list `user_data['watched']`
ratings_sum = sum(movie["rating"] for movie in user_data["watched"])
average_rating = ratings_sum / len(user_data["watched"])
return average_rating
```

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,146 @@
...
+            user_data["watchlist"].remove(movie)
+
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def  get_watched_avg_rating(user_data):
+    if len(user_data["watched"]) == 0:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:39]: I like the function structure overall 👍 I recommend using the truthy/falsy value of `user_data["watched"]` so we can both check if it is empty or pointing to `None` all in one:
```python
if not user_data["watched"]:
    return 0
```

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,146 @@
...
+        if movie["genre"] in genre_frequency:
+            genre_frequency[movie["genre"]] += 1
+        else:
+            genre_frequency[movie["genre"]] = 1
+
+    max_num_for_genre = max(genre_frequency.values())
+
+    for genre, frequency in genre_frequency.items():
+        if frequency == max_num_for_genre:
+            return genre
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:70]: Great use of a frequency map and `max`! If it was important to reduce the number of passes we make over data, we use the `max` function in a slightly different way to get it to return the key containing the genre we want. To do this we pass `genre_frequency` as the iterable and using a second argument `key`:
```python
max_genre = max(genre_frequency, key=genre_frequency.get)
return max_genre
```

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,146 @@
...
+    if len(user_data["watched"]) == 0:
+        return None
+    
+    genre_frequency = {}
+
+    for movie in user_data["watched"]:
+        if movie["genre"] in genre_frequency:
+            genre_frequency[movie["genre"]] += 1
+        else:
+            genre_frequency[movie["genre"]] = 1
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:64]: There's a tradeoff of using a little more space for another variable, but to make the code a little easier to read and reduce how many times we need to access `movie["genre"]`, I might suggest creating a variable before the `if` that holds the result of `movie["genre"]` which we can use on the subsequent lines:
```python
for movie in user_data["watched"]:
    movie_genre = movie["genre"]
    if movie_genre in genre_frequency:
        genre_frequency[movie_genre] += 1
    else:
        genre_frequency[movie_genre] = 1
```

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,146 @@
...
+
+    for genre, frequency in genre_frequency.items():
+        if frequency == max_num_for_genre:
+            return genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_collective_friends_movies(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:76]: Really nice use of a helper function!

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,146 @@
...
+    friends_movies_list = get_collective_friends_movies(user_data)
+
+    for movie in user_data["watched"]:
+        if movie not in friends_movies_list:
+            unique_movies.append(movie)
+
+    return unique_movies
+
+
+def get_friends_unique_watched(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:98]: Nice implementations for this wave! 😄

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,146 @@
...
+            unique_movies.append(movie)
+
+    return unique_movies
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    friends_unique_movies = get_friends_unique_watched(user_data)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:112]: Love the code reuse!

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,146 @@
...
+    available_recommendations = []
+
+    for movie in friends_unique_movies:
+        if movie["host"] in user_data["subscriptions"]:
+            available_recommendations.append(movie)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:117]: Nice algorithm! Another way we could approach filtering the `unique_watched` list is with a list comprehension:

```python
# list comprehension
result = [movie for movie in friends_unique_movies if movie["host"] in user_data["subscriptions"]]
```

This line is a bit long, in practice we would split a statement like this across lines, use some extra variables, or shorten some naming to keep under the PEP8 guide of 79 characters max per line:

```python
# list comprehension
result = [movie for movie in friends_unique_movies 
         if movie["host"] in user_data["subscriptions"]]
```

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,146 @@
...
+
+def get_rec_from_favorites(user_data):
+    friends_movies_list = get_collective_friends_movies(user_data)
+    recommendations = []
+
+    for movie in user_data["favorites"]:
+        if movie not in friends_movies_list:
+            recommendations.append(movie)
+
+    return recommendations
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:144]: I like this approach and that it uses your previous helper function! Another approach could use the result of `get_unique_watched`:
```python
unique_movies = get_unique_watched(user_data)

recommendations = []
for movie in unique_movies:
    if movie in user_data['favorites']:
        recommendations.append(movie)

return recommendations
```

**Review by kelsey-steven-ada — COMMENTED** [review]
🎉 Congrats on completing your first project at Ada! 🎉 I’ve added some suggestions & questions, let me know if there's anything I can clarify.
